File: python/3446_sort_matrix_by_diagonals.py
"""
Problem: 3446 - Sort Matrix by Diagonals  
Description: You are given an n × n square matrix of integers grid. Return the matrix such that .The diagonals in the bottom-left triangle (including the main diagonal) are sorted in non-increasing order.The diagonals in the top-right triangle are sorted in non-decreasing order.  
Date: 2025-08-28
"""

class Solution:
    def sortMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        rows, cols = len(mat), len(mat[0])

        # bottom-left & main diagonals → non-increasing
        for r in range(rows):
            self._sort_diagonal(mat, r, 0, increasing=False)

        # top-right diagonals → non-decreasing
        for c in range(1, cols):
            self._sort_diagonal(mat, 0, c, increasing=True)

        return mat

    def _sort_diagonal(
        self,
        mat: List[List[int]],
        row: int,
        col: int,
        increasing: bool
    ) -> None:
        rows, cols = len(mat), len(mat[0])
        length = min(rows - row, cols - col)

        # Extract diagonal
        diag = [mat[row + i][col + i] for i in range(length)]

        # Sort ascending, then reverse for non-increasing
        diag.sort()
        if not increasing:
            diag.reverse()

        # Write back sorted values
        for i, val in enumerate(diag):
            mat[row + i][col + i] = val

File: cpp/3446_sort_matrix_by_diagonals.cpp
"""
Problem: 3446 - Sort Matrix by Diagonals  
Description: You are given an n × n square matrix of integers grid. Return the matrix such that .The diagonals in the bottom-left triangle (including the main diagonal) are sorted in non-increasing order.The diagonals in the top-right triangle are sorted in non-decreasing order.  
Date: 2025-08-28
"""


class Solution {
public:
    vector<vector<int>> sortMatrix(vector<vector<int>>& mat) {
        int rows = mat.size();
        int cols = mat[0].size();

        // bottom-left & main diagonals → non-increasing
        for (int r = 0; r < rows; ++r) {
            sortDiagonal(mat, r, 0, /*increasing=*/false);
        }
        // top-right diagonals → non-decreasing
        for (int c = 1; c < cols; ++c) {
            sortDiagonal(mat, 0, c, /*increasing=*/true);
        }
        return mat;
    }

private:
    void sortDiagonal(vector<vector<int>>& mat,
                      int row, int col,
                      bool increasing) {
        int rows = mat.size();
        int cols = mat[0].size();
        int len = min(rows - row, cols - col);

        // extract
        vector<int> diag(len);
        for (int i = 0; i < len; ++i) {
            diag[i] = mat[row + i][col + i];
        }

        // sort ascending, then reverse if non-increasing
        sort(diag.begin(), diag.end());
        if (!increasing) {
            reverse(diag.begin(), diag.end());
        }

        // write back
        for (int i = 0; i < len; ++i) {
            mat[row + i][col + i] = diag[i];
        }
    }
};
